src/stack/share.rs

```diff
@@ -0,0 +1,77 @@
+use crate::{nb, SocketAddr, TcpClientStack, TcpFullStack, UdpClientStack, UdpFullStack};
+use core::cell::RefCell;
+
+/// Sharable wrapper for a network stack implementation.
+///
+/// An implementation of the stack traits that can contain (and provide provide
+/// single-threaded shared access to) another stack implementation. A direct
+/// implementation can only be used when owned or with a mutable reference.
+/// This implementation will store another implementation internally, and yield
+/// an arbitrary number of shared references to it, which themselves implement
+/// the stack traits.
+pub struct SharableStack<T> {
+	stack: RefCell<T>,
+}
+
+impl<T> SharableStack<T> {
+	/// Create a new SharedStack that contains and uses some other stack implementation.
+	pub fn new(stack: T) -> Self {
+		SharableStack {
+			stack: RefCell::new(stack),
+		}
+	}
+
+	/// Returns a shared reference to the driver that can be used as a first-class implementation.
+	pub fn acquire(&self) -> SharedStack<T> {
+		SharedStack { stack: &self.stack }
+	}
+}
+
+/// Single-thread shared reference to an internal network stack implementation.
+///
+/// This can only be created by calling [`SharableStack::acquire()`]
+pub struct SharedStack<'a, T> {
+	stack: &'a RefCell<T>,
+}
+
+macro_rules! forward {
+    ($func:ident($($v:ident: $IT:ty),*) -> $T:ty) => {
+        fn $func(&mut self, $($v: $IT),*) -> $T {
+            self.stack.borrow_mut().$func($($v),*)
+        }
+    }
+}
+
+impl<'a, T: UdpClientStack> UdpClientStack for SharedStack<'a, T> {
```

Comment:
I would recommend opting for the `where` clause instead of `T: UdpClientStack` - I find it more readable and easy to follow.